airflow/providers/snowflake/hooks/snowflake.py

```diff
@@ -369,15 +369,19 @@ def run(
         with closing(self.get_conn()) as conn:
             self.set_autocommit(conn, autocommit)
 
-            # SnowflakeCursor does not extend ContextManager, so we have to ignore mypy error here
-            with closing(conn.cursor(DictCursor)) as cur:  # type: ignore[type-var]
```

Comment:
isn't this a breaking change?

Reply:
Correct. This is breaking. I proposed to make Hook breaking change while leaving Operators interface intact (I build the dicationary later). This is the final price (one-time) to pay We already have a breaking change of Snowflake provider and I want to yank 4.0.0 and 4.0.1 and make it even more breaking (it is anyhow broken now for SqlCheck ad SqlColumnCheck and likely other places. See more reasoning here https://github.com/apache/airflow/issues/27976#issuecomment-1331514882

Reply:
BTW. This is precisely what I've done with Databricks last week. Breaking change of Databricks proivider with Hook returning "regular" records, but Operator returning Tuples of "description", "records" -  (this is how Databricks addressed the same problem. See https://airflow.apache.org/docs/apache-airflow-providers-databricks/stable/index.html#breaking-changes

I will describe it in detail when I prepare release notes.

It's not that bad as it might seem. People will still be able to downgrade to 3.* version of provider for mutliple months (and even keep it for as long as they want) - only when they will want to use new features of the provider, they will have to mighrate. I think this is **really** good deal :).  And the whole reason why we separated providers - so that we **can** make breaking changes when we want to move faster. Once we unite the DBApiHook behaviour (and add protection that it does not change lightly - this why all the tests and the  https://github.com/apache/airflow/pull/27962 is for - all kind of lineaage, common tooling, amount of code to maintain for current and future sql providers will be much much better.

Reply:
What if the users prefer using DictCursor with the hook?  Is it really necessary to change that?  It's a nice feature of the snowflake connector that it provides such a cursor (and the flexibility to choose).  Just because other, older connectors don't have this option, why should snowflake conform?

Reply:
from a quick search, seems that it's pretty common.

postgres has it: https://www.psycopg.org/docs/extras.html#dictionary-like-cursor
and mysql: https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursordict.html

i think it's a good, very valuable feature.  a very natural way to retrieve and deal with records. i don't think we should change the behavior of the snowflake hook.  i like the idea of adding some way to control the behavior, and configure it from connections.

maybe in the hooks we just have an optionally implementable cursor factory method that returns cursor given conn.

e.g. 

```
def _cursor_factor(cnx):
     return cnx.cursor(DictCursor)
```

and then optionally in future this could be controlled by hook params or airflow connection, depending on the hook.

aside... just because pep249 says "sequence of sequences" i'm not sure that should mean we (e.g. postgres, mysql, snowflake) should never add more capabilities to what a cursor can do.  it can still be dbapi compatible and also go beyond that initial spec with its functionality, no? like parent / subclass.

Reply:
i guess it's discussed in a faq: https://peps.python.org/pep-0249/#frequently-asked-questions

Maybe the problem is thinking about DbApiHook as though it is 1-1 with the dbapi spec. If that were the case, then it would not have any methods apart from `get_conn` which returns the connection.  Another way to think about it is that the DbApiHook is another layer of abstraction that is not quite the same thing as dbapi 2.0, but which _wraps round_ and _uses_ dbapi-compatible clients in order to do work and add more convenience (and therefore is not strictly bound to dbapi 2.0)

Reply:
No. the thing is that we already use DBApiHook assuming that run method behaves like dbapi.

Reply:
We made some assumptions about the API - they were neither maintained, not documented and they were implicit - for example generic transfer. And they "assumed" that the DBApi Hook behaves like`dbapi`.

So WE were doing the "have cake and eat it too" . On one hand we wanted the hooks to behave in the same way, on the other, we allowed them to behave differently. 

Luckily, with this change and milions of tests around this, that time will be over.

I added the parameter to return dictionary (and single test covering it). Should be enough, because the only difference is the type of returned entities. Looking fortward to the approval :)

Reply:
> from a quick search, seems that it's pretty common.

> postgres has it: https://www.psycopg.org/docs/extras.html#dictionary-like-cursor
> and mysql: https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursordict.html

Cool. So I change my opinion. Let's - as we did now with Snowflake - extend other "DBApuHoook". run methods with return_dictionary. Sounds 1-1 mappable to what postgres, snowflake, mysql does - you need to specifically request dictonaries, they are not returned by default. I am fine with that explanation - and withdraw my suggestion about using  dict_cursor. If many providers did that - so be it, this will become de-facto standard, let's  stick to it and map it to "return_dictionaries" parameter.

> i think it's a good, very valuable feature. a very natural way to retrieve and deal with records. i don't think we should change the behavior of the snowflake hook. i like the idea of adding some way to control the behavior, and configure it from connections.

Added it - and in the next iteration of changes (common.sql 1.4) it might even become "standard" feature for all hooks. I have completely no problem with that as long as we keep consistent behaviour.

> aside... just because pep249 says "sequence of sequences" i'm not sure that should mean we (e.g. postgres, mysql, snowflake) should never add more capabilities to what a cursor can do. it can still be dbapi compatible and also go beyond that initial spec with its functionality, no? like parent / subclass.

Sure. let's add it as optional feature. I am with you. But let's not change default behaviour.


> e.g. for snowflake

> def _cursor_factor(cnx):
>     return cnx.cursor(DictCursor)
> and then optionally in future this could be controlled by hook params or airflow connection, depending on the hook.
:
Yes - in the current implementation I added this.

```

    @contextmanager
    def _get_cursor(self, conn: Any, return_dictionaries: bool):
        cursor = None
        try:
            if return_dictionaries:
                cursor = conn.cursor(DictCursor)
            else:
                cursor = conn.cursor()
            yield cursor
        finally:
            if cursor is not None:
                cursor.close()
```

This can easily be in the future moved to DBApiHoook 1.4 with basic implementation returning regular cursor and can be overridden eventually by child hooks. If you feel like this is a good direction - that would be great next step to do :).

Reply:
BTW. I also added the CHANGELOG entry on how I see the communication about the breaking changes. We will yank 4.0.0/4.0.1 (they has not been released yet in any Airflow and it is out just few days and people reported it broken and downgraded it anyway).  Recent changing extras was breaking anyway so we would not mess around with multiple braking release.  

This is the easiest way to come clean out of that @dstandish - unless you have other ways of handling it now, I am all ears, because both 4.0.1 and 4.0.0 are broken and we NEED to do something NOW.

Reply:
Hey @dstandish (and others - I would love to know what you think) - especially @ephraimbuddy @ashb @jedcunningham  @eladkal @uranusjr  - those involved in making the upcoming 2.5.0 release happen.

I think what we have now is really responding to all comments and providing a good path forwards:

* The operator behaves consistently - same way as before
* The user who already used Hook's run can still bring the old feature easily by adding `return_dictionaries=True` flag
* We open up the way for other hooks to also do a similar - dictionary - return option and we can turn it in 1.4 into a "feature" of DBApiHook that child classes can decide to implement or not.

I think the only thing (correct me if I am wrong @dstandish) is breaking. Breaking the single Snowflake Hook is unfortunate, but I think this is the best way to approach it.

Why? 

We have two options (unfortunately we are in the situation that we have to do something to fix the current "broken" state. Status quo is not an option, because things are broken now and we need to fix them fast - before 2.5.0 release publishing because otherwise, by releasing 2.5.0 we know Snowflake's users will have a buggy version of snowflake provider that just does not work for at least few cases. #27978, #27976 are but two manifestations of that reported even if those problems are not yet exposed by constraints/images of airflow. If we release 2.5.0 with those, the problem will get several magnitude of orders worse.

1) We either do it now and take this (LAST) pain of standardising the last remaining DBAPIHook that behaves differently than all other DBApiHooks (literally) by default  - with a clean "breaking" label and easy way to fix (just add "return_dictionaries" parameter in `run`command to fix it. I strongly feel this is the **right** way of solving the problem.

2) or we will continue to carry the pain and continue breaking things for mutiple users becasue the old setting was literally causing everyone who touched the DBApiHook area breaking things unexpectedly.  I am not even sure how to do it - I spent quite a good deal of thinking on how we could do it differently, and I could not figure a good way which does not break **something** - but maybe someone else can take on the task and would like to attempt it differently. As I said, we need to do something and take action, because status quo is that things are seriously broken if we release 2.5.0 without fixing it.,

This is not academic problem that I imagined. The #27978 and #27976 are a clear example of that. Things got broken not because someone was careless or becasue it could be easily prevented - but because someone wanted to fix and improve things. Unfortunately the historical inconsistencies in the way how our internal code is treated caused that it was either very difficult or impossible to foresee. It's almost like we deliberately set a trap on contributors and maintainers who want to evolve and improve our SQL capabilities so that they have to walk into that trap. And we do not want to fix the trap, we want to keep it by adding even more complexity and handling exceptions.

With this change we effectively end that with single, clean cut. Rather than imposing a "death by thousands cuts" oin our users (and make our maintainers feel guilty of making changes that break things  - we deliberately set a clear cut on where we break and from then on everything is consistent - easier, tested, prevented from making accidental changes and open to further improvements and changes. 

WDYT?